src/algorithms/interfaces/UniqueIDGenSvc.h

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2024 Wouter Deconinck
+
+#pragma once
+
+#include <edm4hep/EventHeaderCollection.h>
+#include <algorithms/service.h>
+#include <bitset>
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+namespace algorithms {
+
+class UniqueIDGenSvc : public Service<UniqueIDGenSvc> {
+public:
+  using event_num_t = decltype(std::declval<edm4hep::EventHeader>().getEventNumber());
+  using run_num_t   = decltype(std::declval<edm4hep::EventHeader>().getRunNumber());
+
+  using seed_t = uint64_t;
+
+private:
+  static constexpr size_t event_num_digits =
+      std::numeric_limits<UniqueIDGenSvc::event_num_t>::digits;
+  static constexpr size_t run_num_digits = std::numeric_limits<UniqueIDGenSvc::run_num_t>::digits;
+  static constexpr size_t seed_digits    = std::numeric_limits<UniqueIDGenSvc::seed_t>::digits;
+  static constexpr size_t name_digits    = std::numeric_limits<size_t>::digits;
+
+public:
+  virtual void init() {}
+
+  // calculate a unique id from name and the event header
+  size_t getUniqueID(const edm4hep::EventHeader& evt_header, const std::string_view& name) const {
+    return getUniqueID(evt_header.getEventNumber(), evt_header.getRunNumber(), name);
+  }
+  // calculate a unique id from name and the first element of collection
+  size_t getUniqueID(const edm4hep::EventHeaderCollection& evt_headers,
+                     const std::string_view& name) const {
+    return getUniqueID(evt_headers.at(0), name);
+  }
+  // calculate a unique id from name and the event and run numbers
+  size_t getUniqueID(const event_num_t evt_num, const run_num_t run_num,
+                     const std::string_view& name) const {
```

Comment:
Not strictly necessary to ingest name here, since can pass that on `init`, not `process`, but this is an attempt at sticking close to the k4FWCore approach. Only difference is that we take a `string_view` instead of a `string` (arguably they will want to move to `string_view` anyway).

Reply:
This comment was for information only. Resolving.